File: src/main/java/gargoyle/ct/ui/impl/CTLocalizableLabel.java
package gargoyle.ct.ui.impl;

import gargoyle.ct.messages.LocaleProvider;
import gargoyle.ct.messages.MessageProvider;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;

public class CTLocalizableLabel extends JLabel {
    private static final long serialVersionUID = 1215028617118416457L;

    public CTLocalizableLabel(@NotNull MessageProvider messages, LocaleProvider provider, String textKey, String toolTipTextKey,
                              int alignment) {
        super(textKey, alignment);
        provider.locale().addPropertyChangeListener(event -> update(messages, textKey, toolTipTextKey));
        update(messages, textKey, toolTipTextKey);
        setIcon(null);
        setHorizontalAlignment(alignment);
    }

    private void update(MessageProvider messages, String textKey, String toolTipTextKey) {
        setText(messages.getMessage(textKey));
        setToolTipText(messages.getMessage(toolTipTextKey));
    }
}

File: app/src/main/kotlin/gargoyle/ct/ui/impl/CTLocalizableLabel.kt
package gargoyle.ct.ui.impl

import gargoyle.ct.util.messages.LocaleProvider
import gargoyle.ct.util.messages.MessageProvider
import gargoyle.ct.util.prop.PropertyObservableDelegate
import java.util.*
import javax.swing.JLabel
import kotlin.reflect.KProperty

class CTLocalizableLabel(
    messages: MessageProvider, provider: LocaleProvider, textKey: String, toolTipTextKey: String,
    alignment: Int
) : JLabel(textKey, alignment) {
    init {
        @Suppress("UNUSED_VARIABLE")
        val providerLocale by PropertyObservableDelegate(provider::locale) { _: KProperty<*>, _: Locale, _: Locale ->
            update(messages, textKey, toolTipTextKey)
        }
        update(messages, textKey, toolTipTextKey)
        icon = null
        horizontalAlignment = alignment
    }

    private fun update(messages: MessageProvider, textKey: String, toolTipTextKey: String) {
        text = messages.getMessage(textKey)
        toolTipText = messages.getMessage(toolTipTextKey)
    }
}
